# *Decision Trees in Python with Scikit-Learn*

### A decision tree is one of most frequently and widely used supervised machine learning algorithms that can perform both regression and classification tasks. The intuition behind the decision tree algorithm is simple, yet also very powerful.
### Introduction Decision Trees are a type of Supervised Machine Learning(that is you explain what the input is and what the corresponding output is in the training data) where the data is continuously split accourding to a certain parameter. The tree can be explained by two entities, namely decicion nodes and leaves.

## Implementing Decision Trees with Python Scikit Learn       
### In this section, we will implement the decision tree algorithm using Python's Scikit-Learn library. In the following examples we'll solve both classification as well as regression problems using the decision tree.

## 1. Decision Tree for Classification

### Importing Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Importing the Dataset         
### Since our file is in CSV format, we will use panda's read_csv method to read our CSV data file. Execute the following script to do so:

In [ ]:
from google.colab import files
upload=files.upload()

Saving bill_authentication.csv to bill_authentication.csv


In [ ]:
dataset=pd.read_csv('bill_authentication.csv')
dataset.head()

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

## Data Analysis

In [ ]:
dataset.shape

(1372, 5)

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Variance  1372 non-null   float64
 1   Skewness  1372 non-null   float64
 2   Curtosis  1372 non-null   float64
 3   Entropy   1372 non-null   float64
 4   Class     1372 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


In [ ]:
dataset.describe()

Variance     Skewness     Curtosis      Entropy        Class
count  1372.000000  1372.000000  1372.000000  1372.000000  1372.000000
mean      0.433735     1.922353     1.397627    -1.191657     0.444606
std       2.842763     5.869047     4.310030     2.101013     0.497103
min      -7.042100   -13.773100    -5.286100    -8.548200     0.000000
25%      -1.773000    -1.708200    -1.574975    -2.413450     0.000000
50%       0.496180     2.319650     0.616630    -0.586650     0.000000
75%       2.821475     6.814625     3.179250     0.394810     1.000000
max       6.824800    12.951600    17.927400     2.449500     1.000000

## Preparing the Data

In [ ]:
x=dataset.drop('Class',axis=1)
y=dataset['Class']

In [ ]:
print(x)
print(y)

      Variance  Skewness  Curtosis  Entropy
0      3.62160   8.66610   -2.8073 -0.44699
1      4.54590   8.16740   -2.4586 -1.46210
2      3.86600  -2.63830    1.9242  0.10645
3      3.45660   9.52280   -4.0112 -3.59440
4      0.32924  -4.45520    4.5718 -0.98880
...        ...       ...       ...      ...
1367   0.40614   1.34920   -1.4501 -0.55949
1368  -1.38870  -4.87730    6.4774  0.34179
1369  -3.75030 -13.45860   17.5932 -2.77710
1370  -3.56370  -8.38270   12.3930 -1.28230
1371  -2.54190  -0.65804    2.6842  1.19520

[1372 rows x 4 columns]
0       0
1       0
2       0
3       0
4       0
       ..
1367    1
1368    1
1369    1
1370    1
1371    1
Name: Class, Length: 1372, dtype: int64


## Splitting Data       
### To understand model performance, dividing the dataset into a training set and a test set is a good strategy.

In [ ]:
# Import train_test_split function
from sklearn.model_selection import train_test_split

In [ ]:
# Split dataset into training set and test set
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20)

## Training and Making Predictions

In [ ]:
# import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

In [ ]:
#create a D-Tree classifier and fit-it
classifier = DecisionTreeClassifier()
classifier.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [ ]:
#Predict the response for test dataset
y_pred = classifier.predict(x_test)
y_pred

array([1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0])

## Evaluating the Model

In [ ]:
#import confusion_matrix
from sklearn.metrics import classification_report, confusion_matrix

In [ ]:
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[156   1]
 [  2 116]]
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       157
           1       0.99      0.98      0.99       118

    accuracy                           0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275



### From the confusion matrix, you can see that out of 275 test instances,            our algorithm misclassified only 4. This is 98.5 % accuracy.

# 2. Decision Tree for Regression
### The process of solving regression problem with decision tree using Scikit Learn is very similar to that of classification. However for regression we use DecisionTreeRegressor class of the tree library. Also the evaluation matrics for regression differ from those of classification. The rest of the process is almost same

## Importing the Dataset

In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
from google.colab import files
upload=files.upload()

Saving petrol_consumption (2).csv to petrol_consumption (2).csv


In [ ]:
dataset=pd.read_csv('petrol_consumption (2).csv')

In [ ]:
dataset.head()

Petrol_tax  Average_income  ...  Population_Driver_licence(%)  Petrol_Consumption
0         9.0            3571  ...                         0.525                 541
1         9.0            4092  ...                         0.572                 524
2         9.0            3865  ...                         0.580                 561
3         7.5            4870  ...                         0.529                 414
4         8.0            4399  ...                         0.544                 410

[5 rows x 5 columns]

### Data Analysis

In [ ]:
dataset.shape

(48, 5)

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Petrol_tax                    48 non-null     float64
 1   Average_income                48 non-null     int64  
 2   Paved_Highways                48 non-null     int64  
 3   Population_Driver_licence(%)  48 non-null     float64
 4   Petrol_Consumption            48 non-null     int64  
dtypes: float64(2), int64(3)
memory usage: 2.0 KB


In [ ]:
dataset.describe()

Petrol_tax  ...  Petrol_Consumption
count   48.000000  ...           48.000000
mean     7.668333  ...          576.770833
std      0.950770  ...          111.885816
min      5.000000  ...          344.000000
25%      7.000000  ...          509.500000
50%      7.500000  ...          568.500000
75%      8.125000  ...          632.750000
max     10.000000  ...          968.000000

[8 rows x 5 columns]

## Preparing the Data

In [ ]:
x=dataset.drop('Petrol_Consumption',axis=1)
y=dataset['Petrol_Consumption']

In [ ]:
print(x)
print(y)

    Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)
0         9.00            3571            1976                         0.525
1         9.00            4092            1250                         0.572
2         9.00            3865            1586                         0.580
3         7.50            4870            2351                         0.529
4         8.00            4399             431                         0.544
5        10.00            5342            1333                         0.571
6         8.00            5319           11868                         0.451
7         8.00            5126            2138                         0.553
8         8.00            4447            8577                         0.529
9         7.00            4512            8507                         0.552
10        8.00            4391            5939                         0.530
11        7.50            5126           14186                         0.525

## Splitting Data         
### To understand model performance, dividing the dataset into a training set and a test set is a good strategy.

In [ ]:
# Import train_test_split function
from sklearn.model_selection import train_test_split

In [ ]:
# Split dataset into training set and test set
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

## Training and Making Predictions

In [ ]:
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor()
regressor.fit(x_train, y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [ ]:
y_pred = regressor.predict(x_test)
y_pred

array([541., 414., 574., 554., 631., 640., 628., 649., 414., 464.])

### Now let's compare some of our predicted values with the actual values and see how accurate we were:

In [37]:
df=pd.DataFrame({'Actual':y_test, 'Predicted':y_pred})
df

Actual  Predicted
29     534      541.0
4      410      414.0
26     577      574.0
30     571      554.0
32     577      631.0
37     704      640.0
34     487      628.0
40     587      649.0
7      467      414.0
10     580      464.0

### Remember that in your case the records compared may be different, depending upon the training and testing split. Since the train_test_split method randomly splits the data we likely won't have the same training and test sets.

# Evaluating the Algorithm
### To evaluate performance of the regression algorithm, the commonly used metrics are mean absolute error, mean squared error, and root mean squared error. The Scikit-Learn library contains functions that can help calculate these values for us. To do so, use this code from the metrics package:

In [38]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 52.1
Mean Squared Error: 4736.5
Root Mean Squared Error: 68.82223477917583


### The mean absolute error for our algorithm is 52.1, which is less than 10 percent of the mean of all the values in the 'Petrol_Consumption' column. This means that our algorithm did a fine prediction job.